include/deal.II/fe/fe_p1nc.h

```diff
@@ -0,0 +1,311 @@
+ *  |                   |
+ *  |                   |
+ *  +--------0.0--------+
+ *  @endverbatim
+ *
+ * </ul>
+ *
+ * Note that above shape functions are constructed on each cell, not on the reference cell only.
+ * @p get_linear_shape computes the coefficients for shape functions when @p fill_fe_values
+ * is called on each cell.
+ *
+ * The basis function associated with a node in global scope is defined by the composition of
+ * (local) basis functions associated with the node on each element.
+ * And DOF associated with the node represents the coefficient of the basis function associated with the node.
+ * In this frame, all DOFs are truly independent.
+ * When a problem with homogeneous Dirichlet boundary condition is considered, the total number of DOFs
+ * is equal to the number of interior nodes, as the number of DOFs which the standard bilinear
+ * finite element @p Q_1 has.
+ *
+ * <h3>Unit support points</h3>
+ * Contrast with ordinary Lagrange finite elements, DOF value with respect to the P1 nonconforming
+ * element at given node does not coincide with the function value at that node.
+ * For instance, the basis function associated with a node has value 0.75 at that node, not 1.0.
+ * Thus we need an interpolation operator which maps any smooth function into a function
+ * with proper DOF values in the P1 element space.
+ * One natural interpolant associated with given smooth function is the linear function whose midpoint
+ * value at each edge is defined by the average of two values at endpoints of the edge.
+ * It provides appropriate weights used in @p unit_support_points.
+ *
+ * <h3>References</h3>
+ * You can find the paper about the P1NC element at
+ * http://epubs.siam.org/doi/abs/10.1137/S0036142902404923.
+ *
+ **/
+
+class FE_P1NC : public FiniteElement<2,2>
+{
+
+public:
+  /**
+   * Constructor for nonparametric version of P1 nonconforming element.
+   */
+  FE_P1NC() ;
+
+  /**
+   * Return the name of the class for the element.
+   */
+  virtual std::string get_name () const ;
+
+  /**
+   * Return the update flags which are needed.
+   */
+  virtual UpdateFlags requires_update_flags (const UpdateFlags flags) const ;
+
+  /**
+   * Copy constructor.
+   */
+  virtual FiniteElement<2,2> *clone () const ;
+
+  /**
+   * Destructor.
+   */
+  virtual ~FE_P1NC ();
+
+
+
+private:
+
+  static std::vector<ComponentMask> get_nonzero_component();
+
+  /**
+   * Return the vector consists of the numbers of degrees of freedom per objects.
+   */
+  static std::vector<unsigned int> get_dpo_vector ();
+
+  /**
+   * Compute the linear shape functions phi(x,y) = ax + by + c
+   * such that each midpoint value on two connecting edges is a half,
+   * and two other midpoint values are all zero.
+   */
+  static void
+  get_linear_shape (const Triangulation<2,2>::cell_iterator &cell,
+                    std::vector<double> &a,
+                    std::vector<double> &b,
+                    std::vector<double> &c);
```

Comment:
This interface is not very efficient because it requires the place where you call the function to allocate 3 `std::vector` objects (requiring memory allocation on the heap). I think a better interface for this function would realize that what the function computes is simply a `3x4` array of doubles, i.e., an array of known sizes. 

It would be useful to also reflect this in the interface of the function. For example, you could write it like this:
```
static
std_cxx11::array<std_cxx11::array<double,3>,4>
get_linear_shape_coefficients (const Tr::cell_iterator &cell);
```